packages/server/src/queries/complex/portfolio/allocation.ts

```diff
@@ -0,0 +1,184 @@
+import { CoinPretty } from "@keplr-wallet/unit";
+import { AssetList } from "@osmosis-labs/types";
+import cachified, { CacheEntry } from "cachified";
+import { LRUCache } from "lru-cache";
+
+import { queryAllocation } from "../../../queries/data-services";
+import { Categories } from "../../../queries/data-services";
+import { AccountCoinsResult } from "../../../queries/data-services";
+import { DEFAULT_LRU_OPTIONS } from "../../../utils/cache";
+import { getAsset } from "../assets";
+
+const allocationCache = new LRUCache<string, CacheEntry>(DEFAULT_LRU_OPTIONS);
+
+interface FormattedAllocation {
+  key: string;
+  percentage: number;
+  amount: number;
+  asset?: CoinPretty;
+}
+
+export interface GetAllocationResponse {
+  all: FormattedAllocation[];
+  assets: FormattedAllocation[];
+  available: FormattedAllocation[];
+}
+
+async function getAll(categories: Categories) {
+  const userBalancesCap = +categories["user-balances"].capitalization;
+  const stakedCap = +categories["staked"].capitalization;
+  const unstakingCap = +categories["unstaking"].capitalization;
+  const unclaimedRewardsCap = +categories["unclaimed-rewards"].capitalization;
+  const pooledCap = +categories["pooled"].capitalization;
+
+  const totalCap =
+    userBalancesCap +
+    stakedCap +
+    unstakingCap +
+    unclaimedRewardsCap +
+    pooledCap;
+
+  return [
+    {
+      key: "available",
+      percentage: (userBalancesCap / totalCap) * 100,
+      amount: userBalancesCap,
+    },
+    {
+      key: "staked",
+      percentage: (stakedCap / totalCap) * 100,
+      amount: stakedCap,
+    },
+    {
+      key: "unstaking",
+      percentage: (unstakingCap / totalCap) * 100,
+      amount: unstakingCap,
+    },
+    {
+      key: "unclaimedRewards",
+      percentage: (unclaimedRewardsCap / totalCap) * 100,
+      amount: unclaimedRewardsCap,
+    },
+    {
+      key: "pooled",
+      percentage: (pooledCap / totalCap) * 100,
+      amount: pooledCap,
+    },
+  ];
+}
+
+async function getAssets(categories: Categories, assetLists: AssetList[]) {
```

Comment:
It seems these functions are highly similar.

Thoughts on only keeping one of these and calling it something like `calcCategoryPercentages`? Then, it could take in another parameter that is "user-balances" or "total-assets" to determine which one to calculate for.